securedrop/journalist_app/utils.py

```diff
@@ -165,10 +166,25 @@ def download(zip_basename: str, submissions: List[Union[Submission, Reply]]) ->
     attachment_filename = "{}--{}.zip".format(
         zip_basename, datetime.datetime.utcnow().strftime("%Y-%m-%d--%H-%M-%S"))
 
-    # Mark the submissions that have been downloaded as such
-    for submission in submissions:
-        submission.downloaded = True
-    db.session.commit()
+    # mark as seen by the current user and update downloaded for submissions
+    for item in submissions:
+        journalist_id = g.get('user').id
+        try:
+            if item.filename.endswith('reply.gpg'):
+                seen_reply = SeenReply(reply_id=item.id, journalist_id=journalist_id)
+                db.session.add(seen_reply)
+            elif item.filename.endswith('-doc.gz.gpg'):
+                seen_file = SeenFile(file_id=item.id, journalist_id=journalist_id)
+                db.session.add(seen_file)
+                item.downloaded = True
+            else:
+                seen_message = SeenMessage(message_id=item.id, journalist_id=journalist_id)
+                db.session.add(seen_message)
+                item.downloaded = True
+
+            db.session.commit()
+        except IntegrityError:
+            pass  # expected not to store that a file was seen by the same user multiple times
```

Comment:
i'm super confused as to why this IntegrityError handling works for single files/messages/replies when i download the same file many times, but somehow this doesn't catch the exception for bulk downloads :/